test/docker-e2e/e2e_state_sync_test.go

```diff
@@ -214,3 +210,423 @@ func (s *CelestiaTestSuite) TestStateSyncMocha() {
 
 	s.Require().NoError(err, "failed to wait for state sync to complete")
 }
+
+// TestStateSyncCompatibilityAcrossUpgrade verifies that a full node can state-sync to the latest height
+// when the chain history includes an on-chain upgrade.
+func (s *CelestiaTestSuite) TestStateSyncCompatibilityAcrossUpgrade() {
+	t := s.T()
+	if testing.Short() {
+		t.Skip("skipping state sync compatibility across upgrade test in short mode")
+	}
+
+	ctx := context.Background()
+
+	tag, err := dockerchain.GetCelestiaTagStrict()
+	s.Require().NoError(err)
+
+	t.Log("Phase 1: Starting v4 chain with 4 validators")
```

Comment:
[nit] v4 will be incorrect when we bump appconsts.Version. Can this log instead use `appconsts.Version - 1`?